File: rust/src/observability/logger.rs
//! Structured logging module for links-queue.
//!
//! Provides structured logging with JSON format support, configurable log levels,
//! and correlation IDs for request tracing.

use std::collections::HashMap;
use std::fmt;
use std::io::{self, Write};
use std::sync::{Arc, RwLock};
use std::time::{SystemTime, UNIX_EPOCH};
use std::cell::RefCell;

/// Log levels with numeric severity.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[repr(u8)]
pub enum LogLevel {
    Debug = 0,
    Info = 1,
    Warn = 2,
    Error = 3,
    Fatal = 4,
}

impl LogLevel {
    /// Returns the string name of the log level.
    #[must_use]
    pub fn as_str(&self) -> &'static str {
        match self {
            LogLevel::Debug => "DEBUG",
            LogLevel::Info => "INFO",
            LogLevel::Warn => "WARN",
            LogLevel::Error => "ERROR",
            LogLevel::Fatal => "FATAL",
        }
    }

    /// Parses a string into a log level.
    #[must_use]
    pub fn from_str(s: &str) -> Option<LogLevel> {
        match s.to_uppercase().as_str() {
            "DEBUG" => Some(LogLevel::Debug),
            "INFO" => Some(LogLevel::Info),
            "WARN" | "WARNING" => Some(LogLevel::Warn),
            "ERROR" => Some(LogLevel::Error),
            "FATAL" => Some(LogLevel::Fatal),
            _ => None,
        }
    }
}

impl fmt::Display for LogLevel {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.as_str())
    }
}

impl Default for LogLevel {
    fn default() -> Self {
        LogLevel::Info
    }
}

/// A structured log entry.
#[derive(Debug, Clone)]
pub struct LogEntry {
    /// Log level.
    pub level: LogLevel,
    /// Log message.
    pub message: String,
    /// Timestamp in milliseconds since Unix epoch.
    pub timestamp: u64,
    /// Optional correlation ID for request tracing.
    pub correlation_id: Option<String>,
    /// Additional fields.
    pub fields: HashMap<String, serde_json::Value>,
}

impl LogEntry {
    /// Creates a new log entry.
    #[must_use]
    pub fn new(level: LogLevel, message: impl Into<String>) -> Self {
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_millis() as u64)
            .unwrap_or(0);

        Self {
            level,
            message: message.into(),
            timestamp,
            correlation_id: LogContext::get_correlation_id(),
            fields: HashMap::new(),
        }
    }

    /// Adds a field to the log entry.
    #[must_use]
    pub fn with_field(mut self, key: impl Into<String>, value: impl Into<serde_json::Value>) -> Self {
        self.fields.insert(key.into(), value.into());
        self
    }

    /// Adds multiple fields to the log entry.
    #[must_use]
    pub fn with_fields(mut self, fields: HashMap<String, serde_json::Value>) -> Self {
        self.fields.extend(fields);
        self
    }

    /// Formats the entry as JSON.
    #[must_use]
    pub fn to_json(&self) -> String {
        let mut obj = serde_json::Map::new();

        obj.insert("timestamp".to_string(), serde_json::json!(self.timestamp));
        obj.insert("level".to_string(), serde_json::json!(self.level.as_str()));
        obj.insert("message".to_string(), serde_json::json!(self.message));

        if let Some(ref cid) = self.correlation_id {
            obj.insert("correlationId".to_string(), serde_json::json!(cid));
        }

        for (key, value) in &self.fields {
            obj.insert(key.clone(), value.clone());
        }

        serde_json::Value::Object(obj).to_string()
    }

    /// Formats the entry as text.
    #[must_use]
    pub fn to_text(&self) -> String {
        let datetime = chrono::DateTime::from_timestamp_millis(self.timestamp as i64)
            .map(|dt| dt.format("%Y-%m-%dT%H:%M:%S%.3fZ").to_string())
            .unwrap_or_else(|| self.timestamp.to_string());

        let mut parts = vec![
            datetime,
            format!("[{}]", self.level),
        ];

        if let Some(ref cid) = self.correlation_id {
            parts.push(format!("[{}]", cid));
        }

        parts.push(self.message.clone());

        // Add extra fields
        let extra: Vec<String> = self.fields
            .iter()
            .filter(|(k, _)| !k.starts_with('_'))
            .map(|(k, v)| format!("{}={}", k, v))
            .collect();

        if !extra.is_empty() {
            parts.push(extra.join(" "));
        }

        parts.join(" ")
    }
}

// Thread-local storage for correlation IDs.
thread_local! {
    static CORRELATION_ID: RefCell<Option<String>> = const { RefCell::new(None) };
}

/// Context manager for correlation IDs.
pub struct LogContext;

impl LogContext {
    /// Gets the current correlation ID.
    #[must_use]
    pub fn get_correlation_id() -> Option<String> {
        CORRELATION_ID.with(|cid| cid.borrow().clone())
    }

    /// Sets the correlation ID for the current thread.
    pub fn set_correlation_id(id: impl Into<String>) {
        CORRELATION_ID.with(|cid| {
            *cid.borrow_mut() = Some(id.into());
        });
    }

    /// Clears the correlation ID for the current thread.
    pub fn clear_correlation_id() {
        CORRELATION_ID.with(|cid| {
            *cid.borrow_mut() = None;
        });
    }

    /// Generates a new UUID correlation ID.
    #[must_use]
    pub fn generate_correlation_id() -> String {
        uuid::Uuid::new_v4().to_string()
    }

    /// Runs a closure with a correlation ID context.
    pub fn run_with<F, R>(correlation_id: impl Into<String>, f: F) -> R
    where
        F: FnOnce() -> R,
    {
        let id = correlation_id.into();
        let previous = Self::get_correlation_id();
        Self::set_correlation_id(id);
        let result = f();
        if let Some(prev) = previous {
            Self::set_correlation_id(prev);
        } else {
            Self::clear_correlation_id();
        }
        result
    }
}

/// Log format type.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LogFormat {
    Json,
    Text,
}

impl Default for LogFormat {
    fn default() -> Self {
        LogFormat::Json
    }
}

/// Output destination for logs.
pub enum LogOutput {
    Stdout,
    Stderr,
    Writer(Box<dyn Write + Send + Sync>),
}

impl Default for LogOutput {
    fn default() -> Self {
        LogOutput::Stdout
    }
}

/// Structured logger for links-queue.
pub struct Logger {
    level: LogLevel,
    format: LogFormat,
    output: Arc<RwLock<LogOutput>>,
    default_fields: HashMap<String, serde_json::Value>,
    include_timestamp: bool,
    include_correlation_id: bool,
}

impl Logger {
    /// Creates a new Logger with default settings.
    #[must_use]
    pub fn new() -> Self {
        Self {
            level: LogLevel::Info,
            format: LogFormat::Json,
            output: Arc::new(RwLock::new(LogOutput::Stdout)),
            default_fields: HashMap::new(),
            include_timestamp: true,
            include_correlation_id: true,
        }
    }

    /// Sets the minimum log level.
    #[must_use]
    pub fn with_level(mut self, level: LogLevel) -> Self {
        self.level = level;
        self
    }

    /// Sets the log format.
    #[must_use]
    pub fn with_format(mut self, format: LogFormat) -> Self {
        self.format = format;
        self
    }

    /// Sets the output destination.
    #[must_use]
    pub fn with_output(mut self, output: LogOutput) -> Self {
        self.output = Arc::new(RwLock::new(output));
        self
    }

    /// Adds a default field.
    #[must_use]
    pub fn with_field(mut self, key: impl Into<String>, value: impl Into<serde_json::Value>) -> Self {
        self.default_fields.insert(key.into(), value.into());
        self
    }

    /// Sets whether to include timestamps.
    #[must_use]
    pub fn with_timestamp(mut self, include: bool) -> Self {
        self.include_timestamp = include;
        self
    }

    /// Sets whether to include correlation IDs.
    #[must_use]
    pub fn with_correlation_id(mut self, include: bool) -> Self {
        self.include_correlation_id = include;
        self
    }

    /// Creates a child logger with additional default fields.
    #[must_use]
    pub fn child(&self, fields: HashMap<String, serde_json::Value>) -> Self {
        let mut new_fields = self.default_fields.clone();
        new_fields.extend(fields);

        Self {
            level: self.level,
            format: self.format,
            output: Arc::clone(&self.output),
            default_fields: new_fields,
            include_timestamp: self.include_timestamp,
            include_correlation_id: self.include_correlation_id,
        }
    }

    /// Gets the current log level.
    #[must_use]
    pub fn level(&self) -> LogLevel {
        self.level
    }

    /// Sets the log level.
    pub fn set_level(&mut self, level: LogLevel) {
        self.level = level;
    }

    /// Checks if a level is enabled.
    #[must_use]
    pub fn is_level_enabled(&self, level: LogLevel) -> bool {
        level >= self.level
    }

    /// Logs a message at the specified level.
    pub fn log(&self, level: LogLevel, message: impl Into<String>, fields: Option<HashMap<String, serde_json::Value>>) {
        if level < self.level {
            return;
        }

        let mut entry = LogEntry::new(level, message);

        // Add default fields
        entry.fields.extend(self.default_fields.clone());

        // Add provided fields
        if let Some(f) = fields {
            entry.fields.extend(f);
        }

        // Handle timestamp inclusion
        if !self.include_timestamp {
            entry.timestamp = 0;
        }

        // Handle correlation ID inclusion
        if !self.include_correlation_id {
            entry.correlation_id = None;
        }

        let output = match self.format {
            LogFormat::Json => entry.to_json(),
            LogFormat::Text => entry.to_text(),
        };

        self.write(&format!("{}\n", output));
    }

    /// Writes output to the destination.
    fn write(&self, data: &str) {
        if let Ok(mut output) = self.output.write() {
            let result = match &mut *output {
                LogOutput::Stdout => io::stdout().write_all(data.as_bytes()),
                LogOutput::Stderr => io::stderr().write_all(data.as_bytes()),
                LogOutput::Writer(w) => w.write_all(data.as_bytes()),
            };
            if let Err(e) = result {
                eprintln!("Logger write error: {}", e);
            }
        }
    }

    /// Logs a debug message.
    pub fn debug(&self, message: impl Into<String>) {
        self.log(LogLevel::Debug, message, None);
    }

    /// Logs a debug message with fields.
    pub fn debug_with(&self, message: impl Into<String>, fields: HashMap<String, serde_json::Value>) {
        self.log(LogLevel::Debug, message, Some(fields));
    }

    /// Logs an info message.
    pub fn info(&self, message: impl Into<String>) {
        self.log(LogLevel::Info, message, None);
    }

    /// Logs an info message with fields.
    pub fn info_with(&self, message: impl Into<String>, fields: HashMap<String, serde_json::Value>) {
        self.log(LogLevel::Info, message, Some(fields));
    }

    /// Logs a warning message.
    pub fn warn(&self, message: impl Into<String>) {
        self.log(LogLevel::Warn, message, None);
    }

    /// Logs a warning message with fields.
    pub fn warn_with(&self, message: impl Into<String>, fields: HashMap<String, serde_json::Value>) {
        self.log(LogLevel::Warn, message, Some(fields));
    }

    /// Logs an error message.
    pub fn error(&self, message: impl Into<String>) {
        self.log(LogLevel::Error, message, None);
    }

    /// Logs an error message with fields.
    pub fn error_with(&self, message: impl Into<String>, fields: HashMap<String, serde_json::Value>) {
        self.log(LogLevel::Error, message, Some(fields));
    }

    /// Logs a fatal message.
    pub fn fatal(&self, message: impl Into<String>) {
        self.log(LogLevel::Fatal, message, None);
    }

    /// Logs a fatal message with fields.
    pub fn fatal_with(&self, message: impl Into<String>, fields: HashMap<String, serde_json::Value>) {
        self.log(LogLevel::Fatal, message, Some(fields));
    }
}

impl Default for Logger {
    fn default() -> Self {
        Self::new()
    }
}

impl Clone for Logger {
    fn clone(&self) -> Self {
        Self {
            level: self.level,
            format: self.format,
            output: Arc::clone(&self.output),
            default_fields: self.default_fields.clone(),
            include_timestamp: self.include_timestamp,
            include_correlation_id: self.include_correlation_id,
        }
    }
}

/// Creates a logger with queue-specific context.
#[must_use]
pub fn create_queue_logger(queue_name: &str, component: Option<&str>) -> Logger {
    let mut logger = Logger::new();
    logger.default_fields.insert("queue".to_string(), serde_json::json!(queue_name));
    if let Some(comp) = component {
        logger.default_fields.insert("component".to_string(), serde_json::json!(comp));
    }
    logger
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_log_level_ordering() {
        assert!(LogLevel::Debug < LogLevel::Info);
        assert!(LogLevel::Info < LogLevel::Warn);
        assert!(LogLevel::Warn < LogLevel::Error);
        assert!(LogLevel::Error < LogLevel::Fatal);
    }

    #[test]
    fn test_log_level_from_str() {
        assert_eq!(LogLevel::from_str("debug"), Some(LogLevel::Debug));
        assert_eq!(LogLevel::from_str("INFO"), Some(LogLevel::Info));
        assert_eq!(LogLevel::from_str("Warning"), Some(LogLevel::Warn));
        assert_eq!(LogLevel::from_str("ERROR"), Some(LogLevel::Error));
        assert_eq!(LogLevel::from_str("fatal"), Some(LogLevel::Fatal));
        assert_eq!(LogLevel::from_str("invalid"), None);
    }

    #[test]
    fn test_log_entry_json() {
        let entry = LogEntry::new(LogLevel::Info, "Test message")
            .with_field("key", "value");

        let json = entry.to_json();
        assert!(json.contains("\"level\":\"INFO\""));
        assert!(json.contains("\"message\":\"Test message\""));
        assert!(json.contains("\"key\":\"value\""));
    }

    #[test]
    fn test_log_entry_text() {
        let entry = LogEntry::new(LogLevel::Info, "Test message");
        let text = entry.to_text();

        assert!(text.contains("[INFO]"));
        assert!(text.contains("Test message"));
    }

    #[test]
    fn test_correlation_id() {
        LogContext::set_correlation_id("test-123");
        assert_eq!(LogContext::get_correlation_id(), Some("test-123".to_string()));

        LogContext::clear_correlation_id();
        assert_eq!(LogContext::get_correlation_id(), None);
    }

    #[test]
    fn test_correlation_id_run_with() {
        LogContext::clear_correlation_id();

        LogContext::run_with("outer-id", || {
            assert_eq!(LogContext::get_correlation_id(), Some("outer-id".to_string()));

            LogContext::run_with("inner-id", || {
                assert_eq!(LogContext::get_correlation_id(), Some("inner-id".to_string()));
            });

            assert_eq!(LogContext::get_correlation_id(), Some("outer-id".to_string()));
        });

        assert_eq!(LogContext::get_correlation_id(), None);
    }

    #[test]
    fn test_logger_level() {
        let logger = Logger::new().with_level(LogLevel::Warn);

        assert!(logger.is_level_enabled(LogLevel::Warn));
        assert!(logger.is_level_enabled(LogLevel::Error));
        assert!(!logger.is_level_enabled(LogLevel::Debug));
        assert!(!logger.is_level_enabled(LogLevel::Info));
    }

    #[test]
    fn test_logger_child() {
        let parent = Logger::new()
            .with_field("parent_field", "parent_value");

        let mut child_fields = HashMap::new();
        child_fields.insert("child_field".to_string(), serde_json::json!("child_value"));
        let child = parent.child(child_fields);

        assert!(child.default_fields.contains_key("parent_field"));
        assert!(child.default_fields.contains_key("child_field"));
    }

    #[test]
    fn test_create_queue_logger() {
        let logger = create_queue_logger("test-queue", Some("worker"));

        assert!(logger.default_fields.contains_key("queue"));
        assert!(logger.default_fields.contains_key("component"));
    }
}

File: js/src/observability/logger.js
/**
 * Structured logging module for links-queue.
 *
 * Provides structured logging with JSON format support, configurable log levels,
 * correlation IDs for request tracing, and log rotation support.
 *
 * @module observability/logger
 *
 * @see REQUIREMENTS.md - REQ-OBS-010 through REQ-OBS-012
 */

import { createWriteStream } from 'node:fs';
import { stat, rename, readdir, unlink } from 'node:fs/promises';
import { dirname, join, basename, extname } from 'node:path';
import { randomUUID } from 'node:crypto';

/**
 * Log levels with numeric severity.
 * Higher numbers indicate more severe levels.
 */
export const LogLevel = Object.freeze({
  DEBUG: 0,
  INFO: 1,
  WARN: 2,
  ERROR: 3,
  FATAL: 4,
});

/**
 * Safely gets an environment variable value.
 * Returns undefined if access is denied (e.g., Deno without --allow-env).
 *
 * @param {string} name - Environment variable name
 * @returns {string|undefined} Value or undefined
 */
const safeGetEnv = (name) => {
  try {
    return process.env[name];
  } catch {
    // Deno without --allow-env will throw
    return undefined;
  }
};

/**
 * Log level names for output.
 */
const LOG_LEVEL_NAMES = {
  [LogLevel.DEBUG]: 'DEBUG',
  [LogLevel.INFO]: 'INFO',
  [LogLevel.WARN]: 'WARN',
  [LogLevel.ERROR]: 'ERROR',
  [LogLevel.FATAL]: 'FATAL',
};

/**
 * Parses a log level string to its numeric value.
 *
 * @param {string|number} level - Log level
 * @returns {number} Numeric log level
 */
const parseLogLevel = (level) => {
  if (typeof level === 'number') {
    return level;
  }
  const upper = String(level).toUpperCase();
  return LogLevel[upper] ?? LogLevel.INFO;
};

/**
 * Formats bytes to human-readable string.
 *
 * @param {number|string} bytes - Bytes or string like "10MB"
 * @returns {number} Bytes
 */
const parseBytes = (bytes) => {
  if (typeof bytes === 'number') {
    return bytes;
  }
  const str = String(bytes).toUpperCase();
  const match = str.match(/^(\d+(?:\.\d+)?)\s*(KB|MB|GB|B)?$/);
  if (!match) {
    return 10 * 1024 * 1024; // Default 10MB
  }
  const value = parseFloat(match[1]);
  const unit = match[2] ?? 'B';
  const multipliers = {
    B: 1,
    KB: 1024,
    MB: 1024 * 1024,
    GB: 1024 * 1024 * 1024,
  };
  return Math.floor(value * (multipliers[unit] || 1));
};

/**
 * Async-local storage for correlation IDs.
 * Falls back to a simple context object if AsyncLocalStorage is not available.
 */
let asyncLocalStorage;
try {
  const { AsyncLocalStorage } = await import('node:async_hooks');
  asyncLocalStorage = new AsyncLocalStorage();
} catch {
  // Fallback for environments without AsyncLocalStorage
  asyncLocalStorage = {
    _store: null,
    getStore() {
      return this._store;
    },
    run(store, fn) {
      const prev = this._store;
      this._store = store;
      try {
        return fn();
      } finally {
        this._store = prev;
      }
    },
    enterWith(store) {
      this._store = store;
    },
  };
}

/**
 * Context manager for correlation IDs.
 */
export class LogContext {
  /**
   * Runs a function with a correlation ID context.
   *
   * @param {string} [correlationId] - Correlation ID (generated if not provided)
   * @param {Function} fn - Function to run
   * @returns {*} Return value of fn
   */
  static run(correlationId, fn) {
    if (typeof correlationId === 'function') {
      fn = correlationId;
      correlationId = randomUUID();
    }
    return asyncLocalStorage.run({ correlationId }, fn);
  }

  /**
   * Runs an async function with a correlation ID context.
   *
   * @param {string} [correlationId] - Correlation ID (generated if not provided)
   * @param {Function} fn - Async function to run
   * @returns {Promise<*>} Return value of fn
   */
  static runAsync(correlationId, fn) {
    if (typeof correlationId === 'function') {
      fn = correlationId;
      correlationId = randomUUID();
    }
    return Promise.resolve(asyncLocalStorage.run({ correlationId }, fn));
  }

  /**
   * Gets the current correlation ID.
   *
   * @returns {string|undefined} Current correlation ID
   */
  static getCorrelationId() {
    return asyncLocalStorage.getStore()?.correlationId;
  }

  /**
   * Sets the correlation ID for the current context.
   *
   * @param {string} correlationId - Correlation ID
   */
  static setCorrelationId(correlationId) {
    const store = asyncLocalStorage.getStore();
    if (store) {
      store.correlationId = correlationId;
    } else {
      asyncLocalStorage.enterWith({ correlationId });
    }
  }

  /**
   * Generates a new correlation ID.
   *
   * @returns {string} New correlation ID
   */
  static generateCorrelationId() {
    return randomUUID();
  }
}

/**
 * Log rotation manager.
 * Handles file rotation based on size or file count.
 */
class LogRotator {
  /**
   * Creates a new LogRotator.
   *
   * @param {Object} options - Rotation options
   * @param {string} options.path - Log file path
   * @param {number|string} [options.maxSize] - Maximum file size before rotation
   * @param {number} [options.maxFiles] - Maximum number of rotated files to keep
   */
  constructor(options) {
    this._path = options.path;
    this._maxSize = parseBytes(options.maxSize ?? '10MB');
    this._maxFiles = options.maxFiles ?? 5;
    this._currentSize = 0;
    this._rotating = false;
  }

  /**
   * Checks if rotation is needed and performs it.
   *
   * @returns {Promise<boolean>} True if rotation was performed
   */
  async checkAndRotate() {
    if (this._rotating) {
      return false;
    }

    try {
      const stats = await stat(this._path);
      this._currentSize = stats.size;

      if (this._currentSize >= this._maxSize) {
        return await this.rotate();
      }
    } catch (error) {
      // File doesn't exist yet, no rotation needed
      if (error.code !== 'ENOENT') {
        console.error('Log rotation check error:', error);
      }
    }
    return false;
  }

  /**
   * Performs log rotation.
   *
   * @returns {Promise<boolean>} True if rotation was successful
   */
  async rotate() {
    if (this._rotating) {
      return false;
    }

    this._rotating = true;
    try {
      const dir = dirname(this._path);
      const ext = extname(this._path);
      const base = basename(this._path, ext);

      // Rotate existing files
      for (let i = this._maxFiles - 1; i >= 1; i--) {
        const oldPath = join(dir, `${base}.${i}${ext}`);
        const newPath = join(dir, `${base}.${i + 1}${ext}`);
        try {
          await rename(oldPath, newPath);
        } catch {
          // File doesn't exist, skip
        }
      }

      // Rename current log to .1
      const rotatedPath = join(dir, `${base}.1${ext}`);
      try {
        await rename(this._path, rotatedPath);
      } catch {
        // Current file doesn't exist or can't be renamed
      }

      // Clean up old files beyond maxFiles
      await this._cleanupOldFiles(dir, base, ext);

      this._currentSize = 0;
      return true;
    } finally {
      this._rotating = false;
    }
  }

  /**
   * Cleans up old rotated log files.
   *
   * @param {string} dir - Directory path
   * @param {string} base - Base filename
   * @param {string} ext - File extension
   * @private
   */
  async _cleanupOldFiles(dir, base, ext) {
    try {
      const files = await readdir(dir);
      const pattern = new RegExp(
        `^${base}\\.(\\d+)${ext.replace('.', '\\.')}$`
      );

      const rotatedFiles = files
        .map((f) => {
          const match = f.match(pattern);
          return match ? { name: f, num: parseInt(match[1], 10) } : null;
        })
        .filter((f) => f !== null)
        .sort((a, b) => b.num - a.num);

      // Remove files beyond maxFiles
      for (const file of rotatedFiles.slice(this._maxFiles)) {
        try {
          await unlink(join(dir, file.name));
        } catch {
          // Ignore deletion errors
        }
      }
    } catch {
      // Ignore errors during cleanup
    }
  }

  /**
   * Updates the current file size.
   *
   * @param {number} bytes - Bytes written
   */
  addBytes(bytes) {
    this._currentSize += bytes;
  }
}

/**
 * JSON log formatter.
 */
class JsonFormatter {
  /**
   * Formats a log entry as JSON.
   *
   * @param {Object} entry - Log entry
   * @returns {string} JSON string
   */
  format(entry) {
    return JSON.stringify(entry);
  }
}

/**
 * Text log formatter.
 */
class TextFormatter {
  /**
   * Formats a log entry as text.
   *
   * @param {Object} entry - Log entry
   * @returns {string} Text string
   */
  format(entry) {
    const { timestamp, level, message, correlationId, ...rest } = entry;
    const parts = [new Date(timestamp).toISOString(), `[${level}]`];

    if (correlationId) {
      parts.push(`[${correlationId}]`);
    }

    parts.push(message);

    // Add extra fields
    const extra = Object.entries(rest)
      .filter(([key]) => !key.startsWith('_'))
      .map(([key, value]) => `${key}=${JSON.stringify(value)}`)
      .join(' ');

    if (extra) {
      parts.push(extra);
    }

    return parts.join(' ');
  }
}

/**
 * Structured logger for links-queue.
 *
 * Supports JSON and text output formats, configurable log levels,
 * correlation ID tracking, and optional file rotation.
 */
export class Logger {
  /**
   * Creates a new Logger.
   *
   * @param {Object} [options] - Logger options
   * @param {string|number} [options.level='info'] - Minimum log level
   * @param {string} [options.format='json'] - Output format ('json' or 'text')
   * @param {string} [options.destination] - File path (defaults to stdout)
   * @param {Object} [options.rotate] - Rotation options
   * @param {string|number} [options.rotate.maxSize] - Maximum file size
   * @param {number} [options.rotate.maxFiles] - Maximum number of rotated files
   * @param {Object} [options.defaultFields] - Default fields to include in all logs
   * @param {boolean} [options.includeTimestamp=true] - Include timestamp in logs
   * @param {boolean} [options.includeCorrelationId=true] - Include correlation ID
   */
  constructor(options = {}) {
    this._level = parseLogLevel(options.level ?? 'info');
    this._format = options.format ?? 'json';
    this._destination = options.destination ?? null;
    this._defaultFields = options.defaultFields ?? {};
    this._includeTimestamp = options.includeTimestamp !== false;
    this._includeCorrelationId = options.includeCorrelationId !== false;

    // Formatter
    this._formatter =
      this._format === 'text' ? new TextFormatter() : new JsonFormatter();

    // File stream and rotation
    this._stream = null;
    this._rotator = null;

    if (this._destination) {
      this._stream = createWriteStream(this._destination, { flags: 'a' });
      if (options.rotate) {
        this._rotator = new LogRotator({
          path: this._destination,
          ...options.rotate,
        });
      }
    }
  }

  /**
   * Gets the current log level.
   *
   * @returns {number} Log level
   */
  get level() {
    return this._level;
  }

  /**
   * Sets the log level.
   *
   * @param {string|number} level - New log level
   */
  setLevel(level) {
    this._level = parseLogLevel(level);
  }

  /**
   * Checks if a level is enabled.
   *
   * @param {number} level - Level to check
   * @returns {boolean} True if enabled
   */
  isLevelEnabled(level) {
    return level >= this._level;
  }

  /**
   * Creates a child logger with additional default fields.
   *
   * @param {Object} fields - Additional default fields
   * @returns {Logger} Child logger
   */
  child(fields) {
    const child = Object.create(this);
    child._defaultFields = { ...this._defaultFields, ...fields };
    return child;
  }

  /**
   * Logs a message at the specified level.
   *
   * @param {number} level - Log level
   * @param {string} message - Log message
   * @param {Object} [fields] - Additional fields
   * @private
   */
  _log(level, message, fields = {}) {
    if (level < this._level) {
      return;
    }

    const entry = {
      ...this._defaultFields,
      level: LOG_LEVEL_NAMES[level],
      message,
      ...fields,
    };

    if (this._includeTimestamp) {
      entry.timestamp = Date.now();
    }

    if (this._includeCorrelationId) {
      const correlationId = LogContext.getCorrelationId();
      if (correlationId) {
        entry.correlationId = correlationId;
      }
    }

    const output = this._formatter.format(entry);
    this._write(`${output}\n`);
  }

  /**
   * Writes output to the destination.
   *
   * @param {string} data - Data to write
   * @private
   */
  _write(data) {
    if (this._stream) {
      this._stream.write(data);
      if (this._rotator) {
        this._rotator.addBytes(Buffer.byteLength(data, 'utf8'));
        // Check rotation asynchronously
        this._rotator.checkAndRotate().catch(() => {});
      }
    } else {
      process.stdout.write(data);
    }
  }

  /**
   * Logs a debug message.
   *
   * @param {string} message - Log message
   * @param {Object} [fields] - Additional fields
   */
  debug(message, fields = {}) {
    this._log(LogLevel.DEBUG, message, fields);
  }

  /**
   * Logs an info message.
   *
   * @param {string} message - Log message
   * @param {Object} [fields] - Additional fields
   */
  info(message, fields = {}) {
    this._log(LogLevel.INFO, message, fields);
  }

  /**
   * Logs a warning message.
   *
   * @param {string} message - Log message
   * @param {Object} [fields] - Additional fields
   */
  warn(message, fields = {}) {
    this._log(LogLevel.WARN, message, fields);
  }

  /**
   * Logs an error message.
   *
   * @param {string} message - Log message
   * @param {Object} [fields] - Additional fields
   */
  error(message, fields = {}) {
    // Handle Error objects
    if (fields instanceof Error) {
      fields = {
        error: fields.message,
        stack: fields.stack,
        name: fields.name,
      };
    }
    this._log(LogLevel.ERROR, message, fields);
  }

  /**
   * Logs a fatal message.
   *
   * @param {string} message - Log message
   * @param {Object} [fields] - Additional fields
   */
  fatal(message, fields = {}) {
    if (fields instanceof Error) {
      fields = {
        error: fields.message,
        stack: fields.stack,
        name: fields.name,
      };
    }
    this._log(LogLevel.FATAL, message, fields);
  }

  /**
   * Flushes the log stream.
   *
   * @returns {Promise<void>}
   */
  flush() {
    if (this._stream) {
      return new Promise((resolve) => {
        this._stream.once('drain', resolve);
        if (this._stream.write('')) {
          resolve();
        }
      });
    }
    return Promise.resolve();
  }

  /**
   * Closes the logger.
   *
   * @returns {Promise<void>}
   */
  close() {
    if (this._stream) {
      return new Promise((resolve, reject) => {
        this._stream.end(() => {
          this._stream = null;
          resolve();
        });
        this._stream.once('error', reject);
      });
    }
    return Promise.resolve();
  }
}

/**
 * Default logger instance.
 */
export const defaultLogger = new Logger({
  level: safeGetEnv('LOG_LEVEL') ?? 'info',
  format: safeGetEnv('LOG_FORMAT') ?? 'json',
});

/**
 * Creates a logger with queue-specific context.
 *
 * @param {Object} options - Logger options
 * @param {string} options.queueName - Queue name
 * @param {string} [options.component] - Component name
 * @returns {Logger} Configured logger
 */
export const createQueueLogger = (options = {}) => {
  const { queueName, component, ...loggerOptions } = options;
  const defaultFields = {};

  if (queueName) {
    defaultFields.queue = queueName;
  }
  if (component) {
    defaultFields.component = component;
  }

  return new Logger({
    ...loggerOptions,
    defaultFields: {
      ...defaultFields,
      ...loggerOptions.defaultFields,
    },
  });
};
